### PR TITLE
EN-3636 Added mobile sdk

### PR DESCRIPTION
**Pull request by pxjohnny**
1) added mobile sdk support
2) fixed tests
3) refactored token/cookie handling (oo)

### CODE REVIEW

**Comment by nitzanpx** [comment]
update readme file regarding the mobile sdk

File: lib/px/block/pxblock.lua [hunk 1]
```diff
@@ -57,62 +70,85 @@ function M.load(config_file)
...
-                        enc_args = ngx_encode_args(req_query_param)
-                        enc_url = ngx_endcode_64(original_req_url .. '?' .. enc_args)
-                    end
-                    local redirect_url = px_config.custom_block_url .. '?url=' .. enc_url .. '&uuid=' .. uuid .. '&vid=' .. vid
-                    px_logger.debug('Redirecting to custom block page: ' .. redirect_url)
-                    ngx_redirect(redirect_url, ngx_HTTP_TEMPORARY_REDIRECT)
+                    collectorUrlBase = 'http://sapi-'
                 end
-            else
+                local collectorUrl = collectorUrlBase .. string.lower(px_config.px_appId) .. '.perimeterx.net'
```
Comment (nitzanpx) [inline lib/px/block/pxblock.lua]: this should be a configurable key in the config

Comment (nitzanpx) [inline lib/px/block/pxblock.lua]: the config should contain certain derivative, baseUrl which its default value is sapi-<app_id>.perimeterx.net and can be override 
Here it should only be set to https/http (that more or like what I think)

File: lib/px/pxnginx.lua [hunk 2]
```diff
@@ -142,65 +142,32 @@ function M.application(file_name)
...
-                -- score did not cross the blocking threshold
-            else
-                ngx.ctx.pass_reason = 'cookie'
-                px_client.send_to_perimeterx("page_requested", details)
-                return true
-            end
-            -- cookie verification failed/cookie does not exist. performing s2s query
-        elseif enable_server_calls == true then
-            return perform_s2s(result, details)
+    px_payload:load(config_file)
```
Comment (nitzanpx) [inline lib/px/pxnginx.lua]: why not pass px_config to the constructor?

File: lib/px/block/pxblock.lua [hunk 3]
```diff
@@ -57,62 +70,85 @@ function M.load(config_file)
...
-                        body = res.body
-                        if ngx.ctx.px_action == 'c' then
-                            body = string_gsub(res.body, '</head>', inject_captcha_script(vid, uuid) .. '</head>', 1);
-                            body = string_gsub(body, '::BLOCK_REF::', uuid);
-                        end
-                    end
-                    ngx.status = ngx_HTTP_FORBIDDEN;
-                    ngx_say(body);
-                    ngx_exit(ngx.OK);
+            if ngx.ctx.px_cookie_origin == "header" then
```
Comment (yaronpx) [inline lib/px/block/pxblock.lua]: add comments that clarify which scenario is each branch

File: lib/px/utils/pxcookiev1.lua [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+function PXCookieV1:process()
+    cookie = ngx.ctx.px_orig_cookie
+    if not cookie then
+        error({ message = "no_cookie" })
+    end
+
+    -- Decrypt AES-256 or base64 decode cookie
+    local data
+
```
Comment (nitzanpx) [inline lib/px/utils/pxcookiev1.lua]: remove extra whitespace :)

File: lib/px/utils/pxcookiev1.lua [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    if fields.u then
+        ngx.ctx.uuid = fields.u
+    end
+
+    if fields.v then
+        ngx.ctx.vid = fields.v
+    end
+
+    -- cookie expired
```
Comment (nitzanpx) [inline lib/px/utils/pxcookiev1.lua]: from here to line 120 this code is duplicated between cookie/tokev v1/3
I think we can take this logic and put in nginx.lua at line 157(ish)
we can discuss about it

File: lib/px/utils/pxpayload.lua [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    -- localized modules
+    self.cjson = require "cjson"
+    self.aes = require "resty.nettle.aes"
+    self.pbkdf2 = require "resty.nettle.pbkdf2"
+    self.hmac = require "resty.nettle.hmac"
+end
+
+function PXPayload:get_payload()
+    ngx.ctx.px_cookie_origin = "cookie"
+    if (ngx.req.get_headers()['X-PX-AUTHORIZATION']) then
```
Comment (nitzanpx) [inline lib/px/utils/pxpayload.lua]: store it in var

**Comment by nitzanpx** [comment]
must have test for token, json respone and 403 status when receiving the response

**Pull request merged**